tower-buffer/src/lib.rs

```diff
@@ -264,15 +277,19 @@ where
 {
     type Response = T::Response;
     type Error = Error<T::Error>;
-    type Future = ResponseFuture<T::Future>;
+    type Future = ResponseFuture<T::Future, T::Error>;
 
     fn poll_ready(&mut self) -> Poll<(), Self::Error> {
         // If the inner service has errored, then we error here.
-        if !self.state.open.load(Ordering::Acquire) {
-            return Err(Error::Closed);
-        } else {
-            self.tx.poll_ready().map_err(|_| Error::Closed)
-        }
+        self.tx.poll_ready().map_err(move |_| {
+            Error::Closed(
+                self.state
```

Comment:
Nit: this code for cloning the error seems repeated between `poll_ready` and `call`, maybe it should be factored out into a separate function?